bss/db/history_db.go

```diff
@@ -0,0 +1,133 @@
+package db
+
+import (
+	"encoding/json"
+	"io/ioutil"
+	"os"
+
+	"github.com/ethereum/go-ethereum/common"
+)
+
+type L2BlockEntry struct {
+	BlockNumber uint64      `json:"block_number"`
+	BlockHash   common.Hash `json:"block_hash"`
+}
+
+type History struct {
+	Entries []L2BlockEntry `json:"entries"`
+}
+
+func (h *History) AppendEntry(entry L2BlockEntry, maxDepth uint64) {
+	var minBlockNumber uint64
+	if entry.BlockNumber > maxDepth {
+		minBlockNumber = entry.BlockNumber - maxDepth
+	}
+
+	var entries = make([]L2BlockEntry, 0, len(h.Entries))
+	for _, e := range h.Entries {
+		// Remove entries that are behind our lookback.
+		if e.BlockNumber < minBlockNumber {
+			continue
+		}
+
+		// Prune entries that occur after the new entry.
+		if e.BlockNumber >= entry.BlockNumber {
```

Comment:
This may break during reorgs I think. Isn't the idea to just keep a history of the last N submitted blocks, regardless of block number? Then the opnode can receive that history (in order from new to old), and find the first common canonical block, to then continue batching from.